app/backend/src/couchers/servicers/search.py

```diff
@@ -0,0 +1,423 @@
+from sqlalchemy.sql import func, or_, text
+
+from couchers import errors
+from couchers.db import session_scope
+from couchers.models import Cluster, FriendRelationship, Page, PageType, PageVersion, User
+from couchers.servicers.api import (
+    hostingstatus2sql,
+    parkingdetails2sql,
+    sleepingarrangement2sql,
+    smokinglocation2sql,
+    user_model_to_pb,
+)
+from couchers.servicers.communities import community_to_pb
+from couchers.servicers.groups import group_to_pb
+from couchers.servicers.pages import page_to_pb
+from couchers.utils import create_coordinate, to_aware_datetime
+from pb import search_pb2, search_pb2_grpc
+
+# searches are a bit expensive, we'd rather send back a bunch of results at once than lots of small pages
+MAX_PAGINATION_LENGTH = 50
+
+regconfig = "english"
+TRI_SIMILARITY_THRESHOLD = 0.6
+TRI_SIMILARITY_WEIGHT = 5
+
+
+def _join_with_space(coalesces):
+    if not coalesces:
+        return ""
+    out = coalesces[0]
+    for coalesce in coalesces[1:]:
+        out += " " + coalesce
+    return out
+
+
+def _build_tsv(A, B=[], C=[], D=[]):
+    """
+    Given lists for A, B, C, and D, builds a tsvector from them.
+    """
+    tsv = func.setweight(func.to_tsvector(regconfig, _join_with_space([func.coalesce(bit, "") for bit in A])), "A")
+    if B:
+        tsv = tsv.concat(
+            func.setweight(func.to_tsvector(regconfig, _join_with_space([func.coalesce(bit, "") for bit in B])), "B")
+        )
+    if C:
+        tsv = tsv.concat(
+            func.setweight(func.to_tsvector(regconfig, _join_with_space([func.coalesce(bit, "") for bit in C])), "C")
+        )
+    if D:
+        tsv = tsv.concat(
+            func.setweight(func.to_tsvector(regconfig, _join_with_space([func.coalesce(bit, "") for bit in D])), "D")
+        )
+    return tsv
+
+
+def _build_doc(A, B=[], C=[], D=[]):
+    """
+    Builds the raw document (without to_tsvector and weighting), used for extracting snippet
+    """
+    doc = _join_with_space([func.coalesce(bit, "") for bit in A])
```

Comment:
Reduce to a oneliner:

`return " ".join(func.coalesce(bit, "") for bit in A + B + C + D)`

Reply:
There's some underlying SQLAlchemy magic going on on top of this, therefore the ugliness with `_join_with_space`, etc.

I had issues with `str.join`: it's expecting what it's joining to be strings, and they're not. I was disappointed to see also that `sum([A,B]) != A + B`...

Reply:
I've added a comment in code

Reply:
`sum()` have a second argument to start off the summation. Needed when concatenating lists like `sum(list_of_lists, [])`